### PR TITLE
Metalness workflow

### PR DESCRIPTION
**Pull request by guycalledfrank**
Support for metalness map instead of colored specular. Specular color is derived from albedo map when using metalness workflow. Specularity is fixed to 0.04 (generic dielectric value) for non-metallic surfaces. New API:
- material.useMetalness (boolean)
- material.metalness (float) is const value or metalness map multiplier
- material.metalnessMap is used as metalness map when metalnessWorkflow is true
- material.metalnessMapTint and obviously tiling/offset/channel/uv

test project:
http://playcanvas.com/designer/339409/scene/355905

sample code:

            metalnessConstMat = app.root.findByName("SphereMC").model.material;
            metalnessConstMat.useMetalness = true;
            metalnessConstMat.metalness = 1;
            metalnessConstMat.update();
            
            var box = app.root.findByName("Box")
            metalnessTexMat = box.model.material;
            metalnessTexMat.useMetalness = true;
            metalnessTexMat.metalness = 1;
            metalnessTexMat.metalnessMap = metalnessTexMat.diffuseMap;
            metalnessTexMat.metalnessMapChannel = "a";
            metalnessTexMat.update();

### CODE REVIEW

**Comment by vkalpias** [comment]
Do we definitely need a public boolean called useMetalness? Couldn't we just make 'metalness' and 'metalnessMap' properties that set an internal boolean (perhaps called _useMetalness) to true when the properties are set?

File: src/graphics/programlib/programlib_phong.js [hunk 1]
```diff
@@ -277,7 +277,7 @@ pc.programlib.phong = {
             } else {
                 code += chunks.specularAaNonePS;
             }
-            code += this._addMap("specular", options, chunks, uvOffset);
+            code += this._addMap(options.metalnessWorkflow? "metalness" : "specular", options, chunks, uvOffset);
```
Comment (vkalpias) [inline src/graphics/programlib/programlib_phong.js]: how about we change options.metalnessWorkflow to options.useMetalness ?

Comment (guycalledfrank) [inline src/graphics/programlib/programlib_phong.js]: "internal boolean" is not intuitive.
metalnessWorkflow is internal, useMetalness is public.

Comment (vkalpias) [inline src/graphics/programlib/programlib_phong.js]: The internal boolean is not meant to be used by the public. They don't really need to set it to true anyway without setting metalness or metalness map. Why not just hide it from the user? It's kinda like the 'blendType' property in scene_material.js which sets whatever booleans it needs internally in the property.

Comment (vkalpias) [inline src/graphics/programlib/programlib_phong.js]: Regarding internal properties like metalnessWorkflow vs useMetalness, is there a reason why they should be named differently? Isn't it more consistent to use the same names for option variables and their counterparts?

Comment (guycalledfrank) [inline src/graphics/programlib/programlib_phong.js]: What I think is not intuitive is metalness overriding specular. Metalness is not necessarily better, but it's different. What happens if you set both metalness and specular? That won't make sense. We need a switch that'll show either specular or metalness controls, useMetalness is that switch.

I agree on consistency, I'll change it.

**Comment by willeastcott** [comment]
Currently, you can set a sphere map or a cube map as the material's reflection map. The cube map will override the sphere map if both are present. Shouldn't we follow that convention here? If a specular map and a metalness map is set, use the metalness map?

**Comment by guycalledfrank** [comment]
Not sure. Cube map is better than a sphere map. Metalness map is no better, but different from a specular map.

**Comment by willeastcott** [comment]
What's the UI for this going to be in the Material Editor (to switch between Specular and Metalness)? @vkalpias @Maksims ?

**Comment by guycalledfrank** [comment]
I thought about checkbox (metalness workflow) that would hide specular map/tint/color controls and show metalness map/slider instead.

**Comment by vkalpias** [comment]
I would guess either a checkbox or a combo box that selects between the two workflows.

**Comment by willeastcott** [comment]
I just imagine that the API would be smart enough to see that:

material.specularMap = someMap;
material.metalnessMap = null;

....is saying that I want to use the specular workflow.

Or even:

var m = new pc.PhongMaterial();
m.metalnessMap = someMap;

Adding m.useMetalness seems a bit....redundant. Sure, a toggle for UI is OK. But is this really necessary for the API? And in the (rare) case where both maps are set, just pick one. Metalness, I'd say, since it's probably less common for that map to be set. Thoughts?

**Comment by guycalledfrank** [comment]
Metalness can be set without any map with just a single float. So you have to have a switch anyway.

**Pull request merged**